src/contracts/client/mailboxclient.cairo

```diff
@@ -0,0 +1,122 @@
+    #[abi(embed_v0)]
+    impl OwnableImpl = OwnableComponent::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+
+    #[storage]
+    struct Storage {
+        mailbox: ContractAddress,
+        local_domain: u32,
+        hook: ContractAddress,
+        interchain_security_module: ContractAddress,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+    }
+
+
+    #[constructor]
+    fn constructor(ref self: ContractState, _mailbox: ContractAddress, _owner: ContractAddress) {
+        self.mailbox.write(_mailbox);
+        let mailbox = IMailboxDispatcher { contract_address: _mailbox };
+        let local_domain = mailbox.get_local_domain();
+        self.local_domain.write(local_domain);
+        self.ownable.initializer(_owner);
+    }
+    #[abi(embed_v0)]
+    impl Upgradeable of IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+            self.ownable.assert_only_owner();
+            self.upgradeable._upgrade(new_class_hash);
+        }
+    }
+
+
+    #[abi(embed_v0)]
+    impl IMailboxClientImpl of IMailboxClient<ContractState> {
+        fn set_hook(ref self: ContractState, _hook: ContractAddress) {
+            self.ownable.assert_only_owner();
+            self.hook.write(_hook);
+        }
+
+        fn set_interchain_security_module(ref self: ContractState, _module: ContractAddress) {
+            self.ownable.assert_only_owner();
+            self.interchain_security_module.write(_module);
+        }
+
+        fn _MailboxClient_initialize(
+            ref self: ContractState,
+            _hook: ContractAddress,
+            _interchain_security_module: ContractAddress,
+        ) {
+            self.ownable.assert_only_owner();
+            self.set_hook(_hook);
+            self.set_interchain_security_module(_interchain_security_module);
+        }
+
+        fn _is_latest_dispatched(self: @ContractState, _id: u256) -> bool {
+            let mailbox_address = self.mailbox.read();
+            let mailbox = IMailboxDispatcher { contract_address: mailbox_address };
+            mailbox.get_latest_dispatched_id() == _id
+        }
+
+        fn _is_delivered(self: @ContractState, _id: u256) -> bool {
+            let mailbox_address = self.mailbox.read();
+            let mailbox = IMailboxDispatcher { contract_address: mailbox_address };
+            mailbox.delivered(_id)
+        }
+
+        fn _dispatch(
+            self: @ContractState,
```

Comment:
I think that's a mutable function so it should be `ref self: ContractState`

Reply:
Also the solidity impl returns `bytes32` please explain the reasoning behind using `u256` in comments

Reply:
We are modifying the storage of the underlying mailbox contract, not the mailboxclient one, that's why. In solidity, the message formatting returns a bytes32, as a result of the keccak hashing process. In our case, we are working with the same keccak hashing returns a u256, so we had to make some changes. I will write a small article with all the significant changes like this one and detail why I considered these changes.